Add an option to DDP to take a list of parameters to ignore upfront.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44826 Add an option to DDP to take a list of parameters to ignore upfront.**

 As described in https://github.com/pytorch/pytorch/issues/43690, there
is a need for DDP to be able to ignore certain parameters in the module (not
install allreduce hooks) for certain use cases. `find_unused_parameters` is
sufficient from a correctness perspective, but we can get better performance
with this upfront list if users know which params are unused, since we won't
have to traverse the autograd graph every iteration.

For now, we have a private static method in DDP to set an attribute for parameters and buffers to be ignored by DDP. The final version of this API still requires additional discussion. 

Differential Revision: [D23740639](https://our.internmc.facebook.com/intern/diff/D23740639/)

rohan-varma committed:
Add an option to DDP to take a list of parameters to ignore upfront.

As described in https://github.com/pytorch/pytorch/issues/43690, there
is a need for DDP to be able to ignore certain parameters in the module (not
install allreduce hooks) for certain use cases. `find_unused_parameters` is
sufficient from a correctness perspective, but we can get better performance
with this upfront list if users know which params are unused, since we won't
have to traverse the autograd graph every iteration.

To enable this, we add a field `parameters_to_ignore` to DDP init and don't
pass in that parameter to reducer if that parameter is in the given list.

Differential Revision: [D23740639](https://our.internmc.facebook.com/intern/diff/D23740639/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add an option to DDP to take a list of parameters to ignore upfront."


Closes https://github.com/pytorch/pytorch/issues/43690. As described in https://github.com/pytorch/pytorch/issues/43690, there
is a need for DDP to be able to ignore certain parameters in the module (not
install allreduce hooks) for certain use cases. `find_unused_parameters` is
sufficient from a correctness perspective, but we can get better performance
with this upfront list if users know which params are unused, since we won't
have to traverse the autograd graph every iteration.

To enable this, we add a field `parameters_to_ignore` to DDP init and don't
pass in that parameter to reducer if that parameter is in the given list.

Differential Revision: [D23740639](https://our.internmc.facebook.com/intern/diff/D23740639/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add an option to DDP to take a list of parameters to ignore upfront."


Closes https://github.com/pytorch/pytorch/issues/43690. As described in https://github.com/pytorch/pytorch/issues/43690, there
is a need for DDP to be able to ignore certain parameters in the module (not
install allreduce hooks) for certain use cases. `find_unused_parameters` is
sufficient from a correctness perspective, but we can get better performance
with this upfront list if users know which params are unused, since we won't
have to traverse the autograd graph every iteration.

To enable this, we add a field `parameters_to_ignore` to DDP init and don't
pass in that parameter to reducer if that parameter is in the given list.

Differential Revision: [D23740639](https://our.internmc.facebook.com/intern/diff/D23740639/)

[ghstack-poisoned]

xw285cornell commented:
Thanks @rohan-varma to work on this! Just have a few comments to discuss: 

* It looks _sync_params_and_buffers uses state_dict which includes parameters and buffers. I need to exclude the buffers in the model parallel part as well because different ranks have different tensors. So I think I need to pass in not only the parameters but also the buffers
* state_dict gives you the tensor instead of nn.Parameter. And it doesn't seem we have very good way to check if two tensors are the same. I see you 're using tensor.data_ptr() to verify, but that may be tricky (two tensor sharing the same storage but different stride? ). I'm wondering if we should change parameters_to_ignore to a dict, with name and tensors. Or maybe just use the fully qualified name instead of parameter/tensors. What do you think?

rohan-varma commented:
@xw285cornell Thanks for taking a look! I agree that `List[str]` from model.named_parameters() used as the comparison makes more sense, as you mentioned offline. In addition to this, I was wondering if this PR should also be updated to supported `broadcast_buffers=True` case.

rohan-varma committed:
Update on "Add an option to DDP to take a list of parameters to ignore upfront."


Closes https://github.com/pytorch/pytorch/issues/43690. As described in https://github.com/pytorch/pytorch/issues/43690, there
is a need for DDP to be able to ignore certain parameters in the module (not
install allreduce hooks) for certain use cases. `find_unused_parameters` is
sufficient from a correctness perspective, but we can get better performance
with this upfront list if users know which params are unused, since we won't
have to traverse the autograd graph every iteration.

To enable this, we add a field `parameters_to_ignore` to DDP init and don't
pass in that parameter to reducer if that parameter is in the given list.

Differential Revision: [D23740639](https://our.internmc.facebook.com/intern/diff/D23740639/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add an option to DDP to take a list of parameters to ignore upfront."


Closes https://github.com/pytorch/pytorch/issues/43690. As described in https://github.com/pytorch/pytorch/issues/43690, there
is a need for DDP to be able to ignore certain parameters in the module (not
install allreduce hooks) for certain use cases. `find_unused_parameters` is
sufficient from a correctness perspective, but we can get better performance
with this upfront list if users know which params are unused, since we won't
have to traverse the autograd graph every iteration.

To enable this, we add a field `parameters_to_ignore` to DDP init and don't
pass in that parameter to reducer if that parameter is in the given list.

Differential Revision: [D23740639](https://our.internmc.facebook.com/intern/diff/D23740639/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add an option to DDP to take a list of parameters to ignore upfront."


Closes https://github.com/pytorch/pytorch/issues/43690. As described in https://github.com/pytorch/pytorch/issues/43690, there
is a need for DDP to be able to ignore certain parameters in the module (not
install allreduce hooks) for certain use cases. `find_unused_parameters` is
sufficient from a correctness perspective, but we can get better performance
with this upfront list if users know which params are unused, since we won't
have to traverse the autograd graph every iteration.

For now, we have a private static method in DDP to set an attribute for parameters and buffers to be ignored by DDP. The final version of this API still requires additional discussion. 

Differential Revision: [D23740639](https://our.internmc.facebook.com/intern/diff/D23740639/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add an option to DDP to take a list of parameters to ignore upfront."


Closes https://github.com/pytorch/pytorch/issues/43690. As described in https://github.com/pytorch/pytorch/issues/43690, there
is a need for DDP to be able to ignore certain parameters in the module (not
install allreduce hooks) for certain use cases. `find_unused_parameters` is
sufficient from a correctness perspective, but we can get better performance
with this upfront list if users know which params are unused, since we won't
have to traverse the autograd graph every iteration.

For now, we have a private static method in DDP to set an attribute for parameters and buffers to be ignored by DDP. The final version of this API still requires additional discussion. 

Differential Revision: [D23740639](https://our.internmc.facebook.com/intern/diff/D23740639/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add an option to DDP to take a list of parameters to ignore upfront."


 As described in https://github.com/pytorch/pytorch/issues/43690, there
is a need for DDP to be able to ignore certain parameters in the module (not
install allreduce hooks) for certain use cases. `find_unused_parameters` is
sufficient from a correctness perspective, but we can get better performance
with this upfront list if users know which params are unused, since we won't
have to traverse the autograd graph every iteration.

For now, we have a private static method in DDP to set an attribute for parameters and buffers to be ignored by DDP. The final version of this API still requires additional discussion. 

Differential Revision: [D23740639](https://our.internmc.facebook.com/intern/diff/D23740639/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add an option to DDP to take a list of parameters to ignore upfront."


 As described in https://github.com/pytorch/pytorch/issues/43690, there
is a need for DDP to be able to ignore certain parameters in the module (not
install allreduce hooks) for certain use cases. `find_unused_parameters` is
sufficient from a correctness perspective, but we can get better performance
with this upfront list if users know which params are unused, since we won't
have to traverse the autograd graph every iteration.

For now, we have a private static method in DDP to set an attribute for parameters and buffers to be ignored by DDP. The final version of this API still requires additional discussion. 

Differential Revision: [D23740639](https://our.internmc.facebook.com/intern/diff/D23740639/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add an option to DDP to take a list of parameters to ignore upfront."


 As described in https://github.com/pytorch/pytorch/issues/43690, there
is a need for DDP to be able to ignore certain parameters in the module (not
install allreduce hooks) for certain use cases. `find_unused_parameters` is
sufficient from a correctness perspective, but we can get better performance
with this upfront list if users know which params are unused, since we won't
have to traverse the autograd graph every iteration.

For now, we have a private static method in DDP to set an attribute for parameters and buffers to be ignored by DDP. The final version of this API still requires additional discussion. 

Differential Revision: [D23740639](https://our.internmc.facebook.com/intern/diff/D23740639/)

[ghstack-poisoned]

mrshenli reviewed (approved):
This LGTM as a temporary solution.
